adam/learner/template_learner.py

```diff
@@ -370,27 +390,24 @@ def _enrich_common(
         def match_template(
             *, concept: Concept, pattern: PerceptionGraphTemplate, score: float
         ) -> None:
-            # try to see if (our model of) its semantics is present in the situation.
-            matcher = pattern.graph_pattern.matcher(
-                preprocessed_perception_graph,
-                match_mode=MatchMode.NON_OBJECT,
-                # debug_callback=self._debug_callback,
+            rtrn = self._match_template(
+                concept=concept,
+                pattern=pattern,
+                perception_graph=preprocessed_perception_graph,
             )
-            for match in matcher.matches(use_lookahead_pruning=True):
-                # if there is a match, use that match to describe the situation.
-                semantic_node_for_match = pattern_match_to_semantic_node(
-                    concept=concept, pattern=pattern, match=match
-                )
+            # Its possible there is no match for the template in the graph
+            # So we handle a None return here
+            if rtrn:
+                (match, semantic_node_for_match) = rtrn
                 match_to_score.append((semantic_node_for_match, score))
                 # We want to replace object matches with their semantic nodes,
                 # but we don't want to alter the graph while matching it,
                 # so we accumulate these to replace later.
                 if isinstance(concept, ObjectConcept):
                     matched_objects.append((semantic_node_for_match, match))
```

Comment:
What's the reason for removing the return here?

Reply:
I moved to this internal `_match_template` method because the Propose but Verify learner model never modifies its internal hypothesis of what a given concept is other than to completely replace it. So there is a modifiable 'confidence' value in how much of the hypothesized pattern overlaps with the actual pattern in the scene. This ability to say "Oh well 90% of the pattern matches so that's ok" means that because our objects are realized as a complex graph structure rather than a single object we can accept some amount of deviance from the object. 

Previously this function required a 100% exact match so by moving to a function where the implementation of the `TemplateLearner` can decide what counts as "a matched template" gives us more flexibility in implementing other learning algorithms.